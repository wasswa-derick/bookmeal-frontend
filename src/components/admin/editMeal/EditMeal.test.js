import React from "react";
import { mount } from "enzyme";
import { EditMeal } from "./EditMeal";

const err = {
  response: {
    status: 400,
    data: {
      errors: {
        form: "simulated errors"
      }
    }
  }
};
const fn = () => Promise.reject(err);
const asyncMockFn = () => Promise.resolve({});

describe("EditMeal", () => {
  let wrapper;
  beforeEach(() => {
    const meal = {
      id: 1,
      title: "",
      description: "",
      price: 0
    };

    const history = {
      push: jest.fn()
    };
    const match = { params: { id: "1" } };

    wrapper = mount(
      <EditMeal
        meal={meal}
        getMeal={asyncMockFn}
        match={match}
        editMeal={fn}
        history={history}
      />
    );
  });

  it("should handle onSubmit", () => {
    wrapper.setState({
      data: {
        id: 1,
        title: "meal title",
        description: "meal desc",
        price: 10000
      }
    });

    const spy = jest.spyOn(wrapper.instance(), "onSubmit");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);

    wrapper
      .find(".btn-save")
      .first()
      .simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should validate data", () => {
    const data = {
      title: "",
      price: "price",
      description: ""
    };
    let errors = wrapper.instance().validate(data);
    expect(errors.title).toEqual("This field is required");
    expect(errors.price).toEqual("Price value should be a number");

    data.price = -1; // change the price to ensure only price greater then zero are accepted
    errors = wrapper.instance().validate(data);
    expect(errors.price).toEqual("Price value cannot be less or equal to zero");
  });

  it("should change state on change of target value", () => {
    const evt = {
      target: {
        name: "title",
        value: "meal title"
      }
    };
    wrapper.instance().onChange(evt);
    const { data } = wrapper.instance().state;
    expect(data.title).toBe("meal title");
  });

  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
