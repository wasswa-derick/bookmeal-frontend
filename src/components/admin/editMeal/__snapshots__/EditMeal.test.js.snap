// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditMeal should render correctly 1`] = `
<EditMeal
  editMeal={[Function]}
  getMeal={[Function]}
  history={
    Object {
      "push": [Function],
    }
  }
  match={
    Object {
      "params": Object {
        "id": "1",
      },
    }
  }
  meal={
    Object {
      "description": "",
      "id": 1,
      "price": 0,
      "title": "",
    }
  }
>
  <div
    className="container"
  >
    <h5>
      Edit Meal: 
    </h5>
    <Component
      component="div"
      loaded={true}
      loadedClassName="loadedContent"
      parentClassName="loader"
    >
      <div
        className="loadedContent"
        key="content"
      >
        <div
          className="modal-dialog"
          role="document"
        >
          <div
            className="modal-content"
          >
            <div
              className="modal-header"
            />
            <div
              className="modal-body"
            >
              <form>
                <FormInput
                  label="Title"
                  name="title"
                  onChange={[Function]}
                  type="text"
                  value=""
                >
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="title"
                    >
                      Title
                    </label>
                    <input
                      autoComplete="off"
                      className="form-control"
                      id="title"
                      name="title"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </div>
                </FormInput>
                <FormInput
                  label="Price"
                  name="price"
                  onChange={[Function]}
                  type="text"
                  value=""
                >
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="price"
                    >
                      Price
                    </label>
                    <input
                      autoComplete="off"
                      className="form-control"
                      id="price"
                      name="price"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </div>
                </FormInput>
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="desc"
                  >
                    Description
                  </label>
                  <textarea
                    className="form-control"
                    cols="20"
                    id="desc"
                    name="description"
                    onChange={[Function]}
                    rows="5"
                    value=""
                  />
                </div>
                <input
                  className="form-control"
                  name="image"
                  type="file"
                />
              </form>
            </div>
            <div
              className="modal-footer"
            >
              <button
                className="btn btn-secondary"
                data-dismiss="modal"
                type="button"
              >
                Close
              </button>
              <button
                className="btn-save btn btn-primary"
                onClick={[Function]}
                type="button"
              >
                Save
              </button>
            </div>
          </div>
        </div>
      </div>
    </Component>
  </div>
</EditMeal>
`;
